sandbox/lua/decoders/json.lua

```diff
@@ -0,0 +1,106 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+Parses a payload containing JSON.
+
+Config:
+
+- type (string, optional, default nil):
+    Sets the message 'Type' header to the specified value
+
+- payload_keep (bool, optional, default false)
+    Always preserve the original log line in the message payload.
+
+.. code-block:: javascript
+
+    {
+      "msg": "Start Request",
+      "event": "artemis.web.ensure-running",
+      "extra": {
+        "workspace-id": "cN907xLngi"
+      },
+      "time": "2015-05-06T20:40:05.509926234Z",
+      "severity": 1
+    }
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [ArtemisLogInput]
+    type = "LogstreamerInput"
+    log_directory = "/srv/artemis/current/logs"
+    file_match = 'artemis\.log'
+    decoder = "JsonDecoder"
+
+    [JsonDecoder]
+    type = "SandboxDecoder"
+    filename = "lua_decoders/json.lua"
+
+        [JsonDecoder.config]
+        type = "artemis"
+        payload_keep = true
+
+*Example Heka Message*
+
+:Timestamp: 2015-05-06 20:40:05 -0800 PST
+:Type: artemis
+:Hostname: test.example.com
+:Pid: 0
+:UUID: 8e414f01-9d7f-4a48-a5e1-ae92e5954df5
+:Payload:
+:EnvVersion:
+:Severity: 1
+:Fields:
+    | name:"msg" value_type:STRING value_string:"Start Request"
+    | name:"event" value_type:STRING value_string:"artemis.web.ensure-running"
+    | name:"extra.workspace-id" value_type:STRING value_string:"cN907xLngi"
+    | name:"time" value_type:STRING value_string:"2015-05-06T20:40:05.509926234Z"
+    | name:"severity" value_type:INTEGER value_integer:1
+--]]
+
+require "cjson"
+local util = require("util")
+
+local msg_type     = read_config("type")
+local payload_keep = read_config("payload_keep")
+
+local msg = {
+    Type       = msg_type,
+    Payload    = nil,
+    Fields     = nil,
+    Severity   = nil
+}
+
+function process_message()
+    local ok, json = pcall(cjson.decode, read_message("Payload"))
+    if not ok then
+        return -1
+    end
+
+    if payload_keep then
+        msg.Payload = read_message("Payload")
+    end
+
+    if json["type"] then
+        msg.Type = json["type"] ; json["type"] = nil
+    end
+
+    if json["severity"] then
+        msg.Severity = json["severity"] ; json["severity"] = nil
+    end
+
+    if json["hostname"] then
+        msg.Hostname = json["hostname"] ; json["hostname"] = nil
+    end
```

Comment:
Rather than hard-coding a check for 'type', 'severity', and 'hostname', I'd like to see a data structure that contains all of the reserved message fields. From the [message schema docs](http://hekad.readthedocs.org/en/v0.10.0b0/message/index.html), the fields I think we want to check for are uuid, timestamp, type, logger, severity, payload, env_version, pid, and hostname. I imagine iterating through the top level keys of the deserialized JSON, lowercasing each value, and checking to see if it's in the set of default fields. If so, try to convert to the right type (timestamp in particular will need to be handled carefully) and then set the value on the msg, removing it from the source table as you're already doing.